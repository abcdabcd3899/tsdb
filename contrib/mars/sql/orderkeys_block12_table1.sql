\set tname test_orderkeys.t1_block12_table1

-- this test is sensitive to cluster size, so force all the data to be stored
-- on seg0 only.
select gp_debug_set_create_table_default_numsegments(1);

-- create :tname with global order (c1), local order (c1, c2)
create table :tname (c1 int, c2 int, id serial)
 using mars
  with (group_param_col_ = "{c1 in 2, id in 100}", local_order_col_= "{c1, c2}")
 distributed by (c1)
;

insert into :tname select i / 4, i % 4 from generate_series(0, 15) i;

\i sql/orderkeys.template
