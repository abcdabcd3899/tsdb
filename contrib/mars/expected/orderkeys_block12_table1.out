\set tname test_orderkeys.t1_block12_table1
-- this test is sensitive to cluster size, so force all the data to be stored
-- on seg0 only.
select gp_debug_set_create_table_default_numsegments(1);
 gp_debug_set_create_table_default_numsegments 
-----------------------------------------------
 1
(1 row)

-- create :tname with global order (c1), local order (c1, c2)
create table :tname (c1 int, c2 int, id serial)
 using mars
  with (group_param_col_ = "{c1 in 2, id in 100}", local_order_col_= "{c1, c2}")
 distributed by (c1)
;
insert into :tname select i / 4, i % 4 from generate_series(0, 15) i;
\i sql/orderkeys.template
--
-- verify the grouping
--
select ctid, c1, c2 from :tname;
 ctid  | c1 | c2 
-------+----+----
 (1,1) |  0 |  0
 (1,2) |  0 |  1
 (1,3) |  0 |  2
 (1,4) |  0 |  3
 (1,5) |  1 |  0
 (1,6) |  1 |  1
 (1,7) |  1 |  2
 (1,8) |  1 |  3
 (2,1) |  2 |  0
 (2,2) |  2 |  1
 (2,3) |  2 |  2
 (2,4) |  2 |  3
 (2,5) |  3 |  0
 (2,6) |  3 |  1
 (2,7) |  3 |  2
 (2,8) |  3 |  3
(16 rows)

--
-- verify the fast seeking with only the first key
--
select c1, c2 from :tname where c1 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
  0 |  3
(4 rows)

select c1, c2 from :tname where c1 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
  0 |  3
  1 |  0
  1 |  1
  1 |  2
  1 |  3
(8 rows)

select c1, c2 from :tname where c1 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
(4 rows)

select c1, c2 from :tname where c1 >= 1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
  2 |  0
  2 |  1
  2 |  2
  2 |  3
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(12 rows)

select c1, c2 from :tname where c1 >  1 order by c1, c2;
 c1 | c2 
----+----
  2 |  0
  2 |  1
  2 |  2
  2 |  3
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(8 rows)

--
-- verify the fast seeking with only the second key
--
select c1, c2 from :tname where c2 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  2 |  0
  3 |  0
(4 rows)

select c1, c2 from :tname where c2 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  1 |  0
  1 |  1
  2 |  0
  2 |  1
  3 |  0
  3 |  1
(8 rows)

select c1, c2 from :tname where c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  1 |  1
  2 |  1
  3 |  1
(4 rows)

select c1, c2 from :tname where c2 >= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  0 |  2
  0 |  3
  1 |  1
  1 |  2
  1 |  3
  2 |  1
  2 |  2
  2 |  3
  3 |  1
  3 |  2
  3 |  3
(12 rows)

select c1, c2 from :tname where c2 >  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  0 |  3
  1 |  2
  1 |  3
  2 |  2
  2 |  3
  3 |  2
  3 |  3
(8 rows)

--
-- verify the fast seeking with both keys
--
select c1, c2 from :tname where c1 <  1 and c2 <  2 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
(2 rows)

select c1, c2 from :tname where c1 <  1 and c2 <= 2 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
(3 rows)

select c1, c2 from :tname where c1 <  1 and c2 =  2 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
(1 row)

select c1, c2 from :tname where c1 <  1 and c2 >= 2 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  0 |  3
(2 rows)

select c1, c2 from :tname where c1 <  1 and c2 >  2 order by c1, c2;
 c1 | c2 
----+----
  0 |  3
(1 row)

select c1, c2 from :tname where c1 <= 1 and c2 <  2 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  1 |  0
  1 |  1
(4 rows)

select c1, c2 from :tname where c1 <= 1 and c2 <= 2 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
  1 |  0
  1 |  1
  1 |  2
(6 rows)

select c1, c2 from :tname where c1 <= 1 and c2 =  2 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  1 |  2
(2 rows)

select c1, c2 from :tname where c1 <= 1 and c2 >= 2 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  0 |  3
  1 |  2
  1 |  3
(4 rows)

select c1, c2 from :tname where c1 <= 1 and c2 >  2 order by c1, c2;
 c1 | c2 
----+----
  0 |  3
  1 |  3
(2 rows)

select c1, c2 from :tname where c1 =  1 and c2 <  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
(2 rows)

select c1, c2 from :tname where c1 =  1 and c2 <= 2 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
(3 rows)

select c1, c2 from :tname where c1 =  1 and c2 =  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  2
(1 row)

select c1, c2 from :tname where c1 =  1 and c2 >= 2 order by c1, c2;
 c1 | c2 
----+----
  1 |  2
  1 |  3
(2 rows)

select c1, c2 from :tname where c1 =  1 and c2 >  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  3
(1 row)

select c1, c2 from :tname where c1 >= 1 and c2 <  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  2 |  0
  2 |  1
  3 |  0
  3 |  1
(6 rows)

select c1, c2 from :tname where c1 >= 1 and c2 <= 2 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  2 |  0
  2 |  1
  2 |  2
  3 |  0
  3 |  1
  3 |  2
(9 rows)

select c1, c2 from :tname where c1 >= 1 and c2 =  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  2
  2 |  2
  3 |  2
(3 rows)

select c1, c2 from :tname where c1 >= 1 and c2 >= 2 order by c1, c2;
 c1 | c2 
----+----
  1 |  2
  1 |  3
  2 |  2
  2 |  3
  3 |  2
  3 |  3
(6 rows)

select c1, c2 from :tname where c1 >= 1 and c2 >  2 order by c1, c2;
 c1 | c2 
----+----
  1 |  3
  2 |  3
  3 |  3
(3 rows)

select c1, c2 from :tname where c1 >  1 and c2 <  2 order by c1, c2;
 c1 | c2 
----+----
  2 |  0
  2 |  1
  3 |  0
  3 |  1
(4 rows)

select c1, c2 from :tname where c1 >  1 and c2 <= 2 order by c1, c2;
 c1 | c2 
----+----
  2 |  0
  2 |  1
  2 |  2
  3 |  0
  3 |  1
  3 |  2
(6 rows)

select c1, c2 from :tname where c1 >  1 and c2 =  2 order by c1, c2;
 c1 | c2 
----+----
  2 |  2
  3 |  2
(2 rows)

select c1, c2 from :tname where c1 >  1 and c2 >= 2 order by c1, c2;
 c1 | c2 
----+----
  2 |  2
  2 |  3
  3 |  2
  3 |  3
(4 rows)

select c1, c2 from :tname where c1 >  1 and c2 >  2 order by c1, c2;
 c1 | c2 
----+----
  2 |  3
  3 |  3
(2 rows)

--
-- verify the fast seeking with AND expr on the first key
--
select c1, c2 from :tname where c1 >= 1 and c1 <= 3 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
  2 |  0
  2 |  1
  2 |  2
  2 |  3
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(12 rows)

select c1, c2 from :tname where c1 >  1 and c1 <= 3 order by c1, c2;
 c1 | c2 
----+----
  2 |  0
  2 |  1
  2 |  2
  2 |  3
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(8 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <  3 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
  2 |  0
  2 |  1
  2 |  2
  2 |  3
(8 rows)

select c1, c2 from :tname where c1 >  1 and c1 <  3 order by c1, c2;
 c1 | c2 
----+----
  2 |  0
  2 |  1
  2 |  2
  2 |  3
(4 rows)

--
-- verify the fast seeking with AND expr on the second key
--
select c1, c2 from :tname where c2 >= 1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  0 |  2
  0 |  3
  1 |  1
  1 |  2
  1 |  3
  2 |  1
  2 |  2
  2 |  3
  3 |  1
  3 |  2
  3 |  3
(12 rows)

select c1, c2 from :tname where c2 >  1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  0 |  3
  1 |  2
  1 |  3
  2 |  2
  2 |  3
  3 |  2
  3 |  3
(8 rows)

select c1, c2 from :tname where c2 >= 1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  0 |  2
  1 |  1
  1 |  2
  2 |  1
  2 |  2
  3 |  1
  3 |  2
(8 rows)

select c1, c2 from :tname where c2 >  1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  1 |  2
  2 |  2
  3 |  2
(4 rows)

--
-- verify the fast seeking with AND expr on both keys
--
select c1, c2 from :tname where c1 >= 1 and c1 <= 3 and c2 >= 1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  1 |  2
  1 |  3
  2 |  1
  2 |  2
  2 |  3
  3 |  1
  3 |  2
  3 |  3
(9 rows)

select c1, c2 from :tname where c1 >  1 and c1 <= 3 and c2 >  1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
  2 |  2
  2 |  3
  3 |  2
  3 |  3
(4 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <  3 and c2 >= 1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  1 |  2
  2 |  1
  2 |  2
(4 rows)

select c1, c2 from :tname where c1 >  1 and c1 <  3 and c2 >  1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
  2 |  2
(1 row)

select c1, c2 from :tname where c1 =  1 and c2 >= 1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  1 |  2
  1 |  3
(3 rows)

select c1, c2 from :tname where c1 =  1 and c2 >  1 and c2 <= 3 order by c1, c2;
 c1 | c2 
----+----
  1 |  2
  1 |  3
(2 rows)

select c1, c2 from :tname where c1 =  1 and c2 >= 1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  1 |  2
(2 rows)

select c1, c2 from :tname where c1 =  1 and c2 >  1 and c2 <  3 order by c1, c2;
 c1 | c2 
----+----
  1 |  2
(1 row)

select c1, c2 from :tname where c1 >= 1 and c1 <= 3 and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  2 |  1
  3 |  1
(3 rows)

select c1, c2 from :tname where c1 >  1 and c1 <= 3 and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  2 |  1
  3 |  1
(2 rows)

select c1, c2 from :tname where c1 >= 1 and c1 <  3 and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  2 |  1
(2 rows)

select c1, c2 from :tname where c1 >  1 and c1 <  3 and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  2 |  1
(1 row)

--
-- verify the IN expr
--
select c1, c2 from :tname where c1 in ( 0, 1, 3 ) order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
  0 |  3
  1 |  0
  1 |  1
  1 |  2
  1 |  3
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(12 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
  0 |  3
(4 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  2
  0 |  3
  1 |  0
  1 |  1
  1 |  2
  1 |  3
(8 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
(4 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 >= 1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(8 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c1 >  1 order by c1, c2;
 c1 | c2 
----+----
  3 |  0
  3 |  1
  3 |  2
  3 |  3
(4 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  3 |  0
(3 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  1 |  0
  1 |  1
  3 |  0
  3 |  1
(6 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  1 |  1
  3 |  1
(3 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 >= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  0 |  2
  0 |  3
  1 |  1
  1 |  2
  1 |  3
  3 |  1
  3 |  2
  3 |  3
(9 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 >  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  2
  0 |  3
  1 |  2
  1 |  3
  3 |  2
  3 |  3
(6 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  3
  1 |  0
  1 |  1
  1 |  3
  2 |  0
  2 |  1
  2 |  3
  3 |  0
  3 |  1
  3 |  3
(12 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  3
(3 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  3
  1 |  0
  1 |  1
  1 |  3
(6 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 =  1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  3
(3 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 >= 1 order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  3
  2 |  0
  2 |  1
  2 |  3
  3 |  0
  3 |  1
  3 |  3
(9 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c1 >  1 order by c1, c2;
 c1 | c2 
----+----
  2 |  0
  2 |  1
  2 |  3
  3 |  0
  3 |  1
  3 |  3
(6 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 <  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  1 |  0
  2 |  0
  3 |  0
(4 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 <= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  1 |  0
  1 |  1
  2 |  0
  2 |  1
  3 |  0
  3 |  1
(8 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 =  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  1 |  1
  2 |  1
  3 |  1
(4 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 >= 1 order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  0 |  3
  1 |  1
  1 |  3
  2 |  1
  2 |  3
  3 |  1
  3 |  3
(8 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 3 ) and c2 >  1 order by c1, c2;
 c1 | c2 
----+----
  0 |  3
  1 |  3
  2 |  3
  3 |  3
(4 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 3 ) and c2 in ( 0, 1, 3 ) order by c1, c2;
 c1 | c2 
----+----
  0 |  0
  0 |  1
  0 |  3
  1 |  0
  1 |  1
  1 |  3
  3 |  0
  3 |  1
  3 |  3
(9 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 2 ) and c1 in ( 1, 2, 3 ) order by c1, c2;
 c1 | c2 
----+----
  1 |  0
  1 |  1
  1 |  2
  1 |  3
  2 |  0
  2 |  1
  2 |  2
  2 |  3
(8 rows)

select c1, c2 from :tname where c2 in ( 0, 1, 2 ) and c2 in ( 1, 2, 3 ) order by c1, c2;
 c1 | c2 
----+----
  0 |  1
  0 |  2
  1 |  1
  1 |  2
  2 |  1
  2 |  2
  3 |  1
  3 |  2
(8 rows)

select c1, c2 from :tname where c1 in ( 0, 1, 2 ) and c1 in ( 1, 2, 3 )
                       and c2 in ( 0, 1, 2 ) and c2 in ( 1, 2, 3 ) order by c1, c2;
 c1 | c2 
----+----
  1 |  1
  1 |  2
  2 |  1
  2 |  2
(4 rows)

